runtime/sema/type.go

```diff
@@ -135,6 +135,10 @@ type Type interface {
 	// IsComparable returns true if values of the type can be compared
 	IsComparable() bool
 
+	// IsMemberAccessible returns true if value of the type can have members values.
+	// Examples are composite types, restricted types, arrays, dictionaries, etc.
+	IsMemberAccessible() bool
```

Comment:
Do we need this notion and `MemberAccessible`? Isn't it possible to access members of all types?

Reply:
This notion is to indicate that a type can be used to access its members using either index-express or member-expression. e.g; `foo.bar` or `foo[bar]`. Only a subset of types has this ability. e.g:
- Composites
- Interfaces
- Arrays (Variable/constant sized)
- Dictionaries
- Restricted types
- Optionals of any of the above.
- 
This is used to determine if the field should be returning a reference or not (primitive fields doesn't need a reference tp be returned)

We could move this to a switch-case (originally that's how I implemented) but it becomes harder to maintain in the long run.

Maybe a better name would avoid any confusions?

Reply:
I see. "Member accessible" in itself just means accessing a member, e.g. `foo.bar`; `foo[bar]` is an indexing access. Maybe rename the type and function to indicate that (maybe `MemberAccessibleOrIndexable`?)

It's still a bit unclear why we need it though, given we already have `ValueIndexableType`/`TypeIndexableType` for indexable types, and `Type.GetMembers` (every type is member-accessible).

Reply:
Yeah, can rename it to `MemberAccessibleOrIndexable`. My first choice was to use `IsContainerType`, but we already use that (and have a `ContainerType`) to represent the notion of "a type that has nested types".

The problem with `ValueIndexableType` is that simple types also implement this, but not all simple types are indexable. Even some indexable types (e.g.: `String`) shouldn't be treated/returned as references.

Reply:
Renamed it to `ContainFieldOrElements` in e47f4517f08f5cb6ea8bf24f8ec8dc88fbc081ed.
Probably not the most appropriate, but hopefully it'll reduce the chance of confusion with "members".

I didn't go with the "ReferenceReturnable" or anything specifying references, because for a type, reference-returnable sounds like "can take a reference to this type", which is not really what we are intending.
The function which calls this method (shouldReturnReference), is the one that actually checks whether a reference should be returned or not, and checking whether the type `ContainFieldOrElements` is only a part of that check.

Reply:
As discussed, we can of-course change this in the future.